kong/clustering/control_plane.lua

```diff
@@ -0,0 +1,534 @@
+local deflate_gzip = utils.deflate_gzip
+local yield = utils.yield
+
+
+local kong_dict = ngx.shared.kong
+local ngx_DEBUG = ngx.DEBUG
+local ngx_NOTICE = ngx.NOTICE
+local ngx_WARN = ngx.WARN
+local ngx_ERR = ngx.ERR
+local ngx_OK = ngx.OK
+local ngx_ERROR = ngx.ERROR
+local ngx_CLOSE = ngx.HTTP_CLOSE
+local PING_INTERVAL = constants.CLUSTERING_PING_INTERVAL
+local PING_WAIT = PING_INTERVAL * 1.5
+local CLUSTERING_SYNC_STATUS = constants.CLUSTERING_SYNC_STATUS
+local DECLARATIVE_EMPTY_CONFIG_HASH = constants.DECLARATIVE_EMPTY_CONFIG_HASH
+local PONG_TYPE = "PONG"
+local RECONFIGURE_TYPE = "RECONFIGURE"
+local _log_prefix = "[clustering] "
+
+
+local function handle_export_deflated_reconfigure_payload(self)
+  ngx_log(ngx_DEBUG, _log_prefix, "exporting config")
+
+  local ok, p_err, err = pcall(self.export_deflated_reconfigure_payload, self)
+  return ok, p_err or err
+end
+
+
+local function is_timeout(err)
+  return err and sub(err, -7) == "timeout"
+end
+
+
+function _M.new(conf, cert_digest)
+  local self = {
+    clients = setmetatable({}, { __mode = "k", }),
+    plugins_map = {},
+
+    conf = conf,
+    cert_digest = cert_digest,
+  }
+
+  return setmetatable(self, _MT)
+end
+
+
+function _M:export_deflated_reconfigure_payload()
+  local config_table, err = declarative.export_config()
+  if not config_table then
+    return nil, err
+  end
+
+  -- update plugins map
+  self.plugins_configured = {}
+  if config_table.plugins then
+    for _, plugin in pairs(config_table.plugins) do
+      self.plugins_configured[plugin.name] = true
+    end
+  end
+
+  -- store serialized plugins map for troubleshooting purposes
+  local shm_key_name = "clustering:cp_plugins_configured:worker_" .. worker_id()
+  kong_dict:set(shm_key_name, cjson_encode(self.plugins_configured))
+  ngx_log(ngx_DEBUG, "plugin configuration map key: " .. shm_key_name .. " configuration: ", kong_dict:get(shm_key_name))
+
+  local config_hash, hashes = calculate_config_hash(config_table)
+
+  local payload = {
+    type = "reconfigure",
+    timestamp = ngx_now(),
+    config_table = config_table,
+    config_hash = config_hash,
+    hashes = hashes,
+  }
+
+  self.reconfigure_payload = payload
+
+  payload, err = cjson_encode(payload)
+  if not payload then
+    return nil, err
+  end
+
+  yield()
+
+  payload, err = deflate_gzip(payload)
+  if not payload then
+    return nil, err
+  end
+
+  yield()
+
+  self.current_hashes = hashes
+  self.current_config_hash = config_hash
+  self.deflated_reconfigure_payload = payload
+
+  return payload, nil, config_hash
+end
+
+
+function _M:push_config()
+  local start = ngx_now()
+
+  local payload, err = self:export_deflated_reconfigure_payload()
+  if not payload then
+    ngx_log(ngx_ERR, _log_prefix, "unable to export config from database: ", err)
+    return
+  end
+
+  local n = 0
+  for _, queue in pairs(self.clients) do
+    table_insert(queue, RECONFIGURE_TYPE)
+    queue.post()
+    n = n + 1
+  end
+
+  ngx_update_time()
```

Comment:
Just a note. We don't have this in 2.8.

Should we check the log level before doing this? `ngx_update_time` is not free.

Reply:
It belongs 3.0.x, I just revert the code and do touch anything.